[WIP] History planes transition

killerducky opened the pull request:
First to curious people, this issue is probably not critical like the move_count. All self-play data, past, present and future is still valid, and we should be able to continue making progress while this issue is being worked on. Simple explanation is the board histories are not oriented the way the Deepmind paper specified, and this makes it a harder for the network to make use of them. But clearly the network is able to play pretty decent chess despite this.

@Error323 @glinscott got a late start on this tonight and only got a tiny bit of work done. 

All I did was create two copies of v1 and v2 data in the old format, and diffed them to verify the game is reproducible. Then I modified the code to send the old format to the network so the game is still identical, but output the new format to the training. Again I ran it twice to make sure the game is the same.

Finally I just glanced at the diff of the v1 old format and the v2 new format. I can see the history planes are different and everything else (policy, value, winner of game) is the same. I haven't looked in detail if the flipping is done correctly yet.

[training.0.gz](https://github.com/glinscott/leela-chess/files/1870247/training.0.gz) - v2 (binary) - use_v1_oldflip=true
[training.4.gz](https://github.com/glinscott/leela-chess/files/1870249/training.4.gz) - v2 (binary) - use_v1_oldflip=false
[training_v1.0.gz](https://github.com/glinscott/leela-chess/files/1870250/training_v1.0.gz) - v1 (text) - use_v1_oldflip=true
[training_v1.4.gz](https://github.com/glinscott/leela-chess/files/1870251/training_v1.4.gz) - v1 (text) - use_v1_oldflip=false

killerducky commented:
Decoded some of this manually. Shoulda just started off with a script cuz it's gonna get confusing fast! I didn't even get far enough to see the flip issue.

```
-----------------------------------
line 1: white's move 1.Nh3 (oldflip=true)
-----------------------------------

00ff000000000000  2nd rank PPPPPPPP   << us pieces
4200000000000000  1st rank -N----N-
2400000000000000  1st rank --B--B--
8100000000000000  1st rank R------R
1000000000000000  1st rank ---Q----
0800000000000000  1st rank ----K---
000000000000ff00  7th rank pppppppp   << them pieces
0000000000000042  8th rank -n----n-
0000000000000024  8th rank --b--b--
0000000000000081  8th rank r------r
0000000000000010  8th rank ---q----
0000000000000008  8th rank ----k---
0000000000000000  rep1=0
0000000000000000  rep2=0
--- cut 7*14 planes of zeros ---
1   black_OOO
1   black_OO
1   white_OOO
1   white_OO
0   us = BLACK
0   rule50_count
0   move_count
--cut policy_pi --
-1  result  (us lose = black win)

-----------------------------------
line 122: black's move 1...Nh6 (oldflip=true)
-----------------------------------
00ff000000000000  2nd rank PPPPPPPP   << us pieces
4200000000000000  1st rank -N----N-
2400000000000000  1st rank --B--B--
8100000000000000  1st rank R------R
1000000000000000  1st rank ---Q----
0800000000000000  1st rank ----K---
000000000000ff00  7th rank pppppppp   << them pieces
0000000000800002  8th rank ------n-   6th rank n-------  (white opened Nh3!)
0000000000000024  8th rank --b--b--
0000000000000081  8th rank r------r
0000000000000010  8th rank ---q----
0000000000000008  8th rank ----k---
0000000000000000  rep1=0
0000000000000000  rep2=0

1 move back in time. Impossible to see any issue because
flipping the opening position vertically gives the identical
position.
00ff000000000000  
4200000000000000
2400000000000000
8100000000000000
1000000000000000
0800000000000000
000000000000ff00
0000000000000042
0000000000000024
0000000000000081
0000000000000010
0000000000000008
0000000000000000
0000000000000000
--- cut 6*14 planes of zeros ---
1   black_OOO
1   black_OO
1   white_OOO
1   white_OO
1   us = BLACK
1   rule50_count
0   move_count
--cut policy_pi --
1  result  (us win = black win)

-----------------------------------
line 243: whites's move 2 (oldflip=true)
-----------------------------------
00ff000000000000
0200800000000000  white Nc1 and Nh3
2400000000000000
8100000000000000
1000000000000000
0800000000000000
000000000000ff00
0000000000010040  black Nh6 and Nc8
0000000000000024
0000000000000081
0000000000000010
0000000000000008
0000000000000000
0000000000000000
```

mooskagh commented:
Probably it makes sense also to fix a black move flip bug?
Here is the permutation table for the policy head indices: https://gist.github.com/mooskagh/9537fe0ea78097334a660b038d72d21f
```
blackMoveToWhiteMove[code('a1b2')] = code('b8c7')
```
It's symmetric, so
```
blackMoveToWhiteMove[blackMoveToWhiteMove[x]] = x
```

killerducky commented:
@mooskagh you're talking about #236 right? Yes I plan to fix that as well.

mooskagh commented:
Yes, I talk about #236.

Error323 commented on training/tf/convert_to_v3.py:
```diff
@@ -0,0 +1,213 @@
+#
+VERSION1 = struct.pack('i', 1)
+VERSION2 = struct.pack('i', 2)
+
```
Minor: You could obtain versions from chunkparser.py

Error323 commented on training/tf/convert_to_v3.py:
```diff
@@ -0,0 +1,213 @@
+        # uint8 move_count (1 byte)
+        # int8 result (1 byte)
+        self.v2_struct = struct.Struct('4s7696s896sBBBBBBBb')
+
```
This function should also not be duplicate?

Error323 commented on training/tf/convert_to_v3.py:
```diff
@@ -0,0 +1,213 @@
+        self.us_black =int(text_item[NUM_HIST*NUM_PLANES+4])
+        self.rule50_count = min(int(text_item[NUM_HIST*NUM_PLANES+5]), 255)  # TODO should be around 99-102ish
+
+    def convert_v1_to_v3(self, ply, text_item):
```
I think it's better to use the chunked here and convert from binary format. chunkparser handles conversion from V1 for free. V1 -> V2 is well tested in chunkparser.

glinscott commented:
We might want to consider normalizing the 50 move counter to a 0..1 range too in the v3 process.

killerducky commented:
@Error323 this is mostly done, it's to the point that we can discuss hand off for the conversion and retrain process. 

Usage:
```
convert_to_v3.py filenames.gz --display
convert_to_v3.py filenames.gz --convert
```
Display pretty prints the training data to stdout.
Convert reads the files and outputs files called `v3bin_OLDFILENAME.gz`. Maybe we can have them go to another directory? Let me know about this.

I made rule50_count a float in chunkparser, is that what we want? It wasn't on a 4 byte boundary so I moved it, does it matter? 

I added `old_movelist.txt` and `new_movelist.txt`. I made them by locally adding some temp code in Network.cpp, because I didn't want to rewrite that logic in python. convert_to_v3.py reads those files to know the order of the policy, and knows how to flip them vertically for black's moves.

TODO list from convert_to_v3.py:
```
# - The chunkparser convert_v3_to_tuple is untested.
#     I just took a guess at it. @Error323 can you help here?
# - Modify lczero code to detect version in weights and switch
#     to V3 automatically
#
# Done:
# - output binary for V3, not text
# - display/flip policy output
# - remove rep2 plane
# - normalize rule50_count to be in [0,1].
# - refactor to use chunckparser.py where appropriate.
#     Still more can be done here but it's closer now.
#
# Plan not to change:
# - remove move_count?
# - I vote to keep side_to_move, but it may not be needed either.
# - LZGo actually added black_to_move white_to_move planes to help
#     it recognize the edge of the board more easily.
```

killerducky commented on src/Network.cpp:
```diff
@@ -474,25 +480,44 @@ void Network::init_move_map() {
+        old_moves.push_back(make<PROMOTION>(from, to, ROOK));
+        old_moves.push_back(make<PROMOTION>(from, to, BISHOP));
+        if (c == BLACK) {
+          // skip underpromotions from 2nd to 1st rank for old
```
It's standard terminology: https://en.wikipedia.org/wiki/Promotion_(chess)

mooskagh replied:
Underpromotions are to rook/bishop/knight, but we store promotions to rook/bishop/queens. Promotion to queen is not an underpromotion.

killerducky replied:
Oh now I see, I'll reword it. As a joke maybe LZ can redefined knight as "promotion" and everything else "underpromotion"? Just kidding.  :)

killerducky commented:
Rebased to latest next.

Error323 commented:
A few remarks:

* rule50_count, normalizing won't work as it was implemented now. Tensorflow space expects input planes to be of type uint8_t. We should perform the normalization in tensorflow space if we want it, see `ChunkParser.parse_function` for more information.
* Current convert_to_v3.py takes 111 hours on my hexacore with hyperthreading for 1M chunks. We need to speed it up.
* I tried to speed it up with pypy but this failed on my system.

I'll begin working on a C++ conversion tool, we can easily check for correctness with the python version. Please review the commits carefully @killerducky. To make sure I didn't do anything stupid.

killerducky commented:
I think it's ready. Fixed a scary bug in the display script, but after fixing looks like the reps=1 stuff is working correctly.

I added comments to the top of convert_v2_v3.cpp about the verification flow comparing the python and c++ conversions. They match!

```
// g++ convert_v2_v3.cpp -std=c++14 -O3 -o convert -lz
//
// Verify against python version:
// ./convert_to_v3.py t_v2bin_oldflip_true.gz --output outpy
// ./convert_to_v3.py outpy/t_v2bin_oldflip_true-v3.gz --display > outpy/t_v2bin_oldflip_true-v3.parse.txt
//
// ./convert t_v2bin_oldflip_true.gz outcpp/cv3bin_t_v2bin_oldflip_true.gz
// ./convert_to_v3.py outcpp/cv3bin_t_v2bin_oldflip_true.gz --display > outcpp/cv3bin_t_v2bin_oldflip_true-parse
// diff outpy/t_v2bin_oldflip_true-v3.parse.txt outcpp/cv3bin_t_v2bin_oldflip_true-parse
```

killerducky commented:
I ran the client with this branch for a single game, @glinscott is there a way to find the training data associated with this game and check it?
http://lczero.org/game/3681435

gsobala commented:
Currently does not compile under darwin, linkage error. Can't figure it out myself. I note `constexpr int Network::T_HISTORY;` has been removed from Network.cpp compared to next and master but it is defined in Network.h so ??

```
undef: __ZN7Network9T_HISTORYE
Undefined symbols for architecture x86_64:
  "Network::T_HISTORY", referenced from:
      Network::gather_features(BoardHistory const&, Network::NNPlanes&) in lto.o
ld: symbol(s) not found for architecture x86_64
```

killerducky commented:
Hm I removed them because it didn't seem necessary. Can you try adding back all the constexpr's from Network.h into Network.cpp and see if it fixes it? If so I can add them back, doesn't hurt anything for my system.

gsobala commented:
Added in `constexpr int Network::T_HISTORY;` into Network.cpp and it compiles and runs. Can't see why just that one declaration should fix it though.

evalon32 commented:
I suspect it's because std::min takes its arguments by reference, and that constitutes ODR-use.
